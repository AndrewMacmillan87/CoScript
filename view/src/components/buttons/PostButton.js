import React from "react";
import PropTypes from "prop-types";

import "./buttons.css";

const PostButton = ({ onClick, classes, value }) => {
    return (
        <input
            onClick={onClick}
            className={classes}
            type='submit'
            value={value}
        />
    );
};

PostButton.defaultProps = {
    value: "Verify"
};

PostButton.propTypes = {
    onClick: PropTypes.func,
    classes: PropTypes.string,
    value: PropTypes.string
};

export default PostButton;
