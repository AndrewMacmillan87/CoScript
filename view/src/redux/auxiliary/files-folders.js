import axios from "axios";

const config = {
    headers: {
        "Content-Type": "application/json",
    },
};

/**
 * Gets all folders that are associated with
 * this project
 *
 * @param {String} projectId - id of the project
 */
export const getAllProjectFolders = async (projectId) => {
    try {
        const res = await axios.get(`/api/folders/${projectId}`);
        return res.data;
    } catch (err) {
        return [];
    }
};

/**
 * Creates a new instance of ProjectFolder in
 * the database and associates it with this
 * project
 *
 * @param {String} projectId - id of the project
 * @param {String} label     - name of the folder
 */
export const createNewFolder = async (projectId, label) => {
    try {
        const body = JSON.stringify({ projectId, label });
        const res = await axios.post("/api/folders/add", body, config);
        return res.data;
    } catch (err) {
        return false;
    }
};

/**
 * Creates a new instance of File in the
 * database and associates it with this project
 *
 * @param {String} folderId  - id of the parent folder
 * @param {String} label     - name of the file
 */
export const createNewFile = async (folderId, label) => {
    try {
        const body = JSON.stringify({ label });
        const res = await axios.post(
            `/api/files/add/${folderId}`,
            body,
            config
        );
        return res.data;
    } catch (err) {
        return false;
    }
};
