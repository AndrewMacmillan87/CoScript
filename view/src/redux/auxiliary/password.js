const axios = require("axios");

/**
 * Compares a newly created password against
 * all items in a user's passwordHistory property
 * in the database
 *
 * @param {Object} user     - logged-in user
 * @param {String} password - new password
 */
export const comparePW = async (user, password) => {
    if (!password) return false;
    try {
        const config = {
            headers: {
                "Content-Type": "application/json",
            },
        };
        const body = JSON.stringify({ user, password });
        const res = await axios.post("/api/changepassword/check", body, config);
        return res.data.clash;
    } catch (err) {
        return true;
    }
};

/**
 * Checks the user's current password to see if it
 * is correct before allowing the user to set a new
 * password
 *
 * @param {Object} user     - logged-in user
 * @param {String} password - user's current password
 */
export const checkCurrentPassword = async (user, password) => {
    if (!password) return false;
    try {
        const config = {
            headers: {
                "Content-Type": "application/json",
            },
        };
        const body = JSON.stringify({ user, password });
        const res = await axios.post(
            "/api/changepassword/check/current",
            body,
            config
        );
        return res.data.correct;
    } catch (err) {
        return false;
    }
};
