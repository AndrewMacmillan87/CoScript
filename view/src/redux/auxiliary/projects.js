/**
 * Shuffles an array of projects randomly, in
 * place, to be displayed in the index page
 * carousel menu
 *
 * @param {Array} array - an array of projects
 */
export const shuffle = (array) => {
    if (array.length > 2) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }
};

/**
 * Breaks up a larger array of projects into a 2D
 * matrix of project array, each of length <width>.
 * Any remaining projects that equate to less than
 * <width> are added at the end of the 2D matrix
 *
 * @param {Array} array  - an array of projects
 * @param {Number} width - number of projects to show
 */
export const segment = (array, width) => {
    const segmentedArray = [];
    if (!array || !width) return segmentedArray;
    let tempArray = [];
    for (let i = 0; i < array.length; i++) {
        tempArray.push(array[i]);
        if (tempArray.length === width) {
            segmentedArray.push(tempArray);
            tempArray = [];
        }
    }
    if (tempArray.length > 0) segmentedArray.push(tempArray);
    return segmentedArray;
};

/**
 * Filters out projects belonging to this user from
 * an array of all projects
 *
 * @param {Array} array   - an array of projects
 * @param {String} userId - id of logged-in user
 */
export const filterUsersProjects = (array, userId) => {
    if (array.length > 0) {
        return array.filter((el) => el.user === userId, []);
    }
};

/**
 * Filters out all active projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterActive = (array) => {
    if (array.length > 0) {
        return array.filter((el) => !el.published, []);
    }
};

/**
 * Filters out all published projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterPublished = (array) => {
    if (array.length > 0) {
        return array.filter((el) => el.published, []);
    }
};

/**
 * Loops over an array of projects to find any projects whose
 * subjects (author, title, genres) contain a matching sub-
 * string to the <searchCriteria>. Returns an array of projects
 * or an empty array if none were matched
 *
 * @param {Array} array           - an array of projects
 * @param {String} status         - the project's published status
 * @param {String} searchCriteria - the search term
 * @param {String} subject        - the field to search in
 */
export const filterByCriteria = (array, status, searchCriteria, subject) => {
    const cleanedSearch = searchCriteria.toLowerCase();
    const results = [];
    for (let project of array) {
        const condition =
            status === "active" ? !project.published : project.published;
        if (
            project[subject].toLowerCase().includes(cleanedSearch) &&
            condition
        ) {
            results.push(project);
        }
    }
    return results;
};
