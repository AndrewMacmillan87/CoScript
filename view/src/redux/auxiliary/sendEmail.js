const axios = require("axios");

const config = {
    headers: {
        "Content-Type": "application/json",
    },
};

/**
 * Sends an email to a user
 *
 * @param {String} email - user's email
 * @param {String} subject - the email subject
 * @param {String} source - process that initiated
 *                          the email being sent
 */
export const sendEmail = async (email, subject, source) => {
    let data;
    await axios
        .post(
            "/api/email/send",
            JSON.stringify({ email, subject, source }),
            config
        )
        .then((result) => (data = result))
        .catch((error) => console.log(error));
    return data;
};

/**
 * Send the user an email with a code they can use to
 * access a section of their account where they can
 * change their password
 *
 * @param {String} email - user's email
 */
export const sendPWEmail = async (email) => {
    let data;
    await axios
        .post("/api/email/send/changepw", JSON.stringify({ email }), config)
        .then((result) => (data = result))
        .catch((error) => console.log(error));
    return data;
};

/**
 * Verify the code a user has entered in order to
 * be granted access to the section of the application
 * that let's them change their password
 *
 * @param {Object} param0 - object containing the codes
 *                          needed to start changing a
 *                          user's password
 */
export const verifyPWCode = async ({
    originalCode,
    confirmCode,
    forgottenDetails,
}) => {
    const res = await axios.post(
        "/api/email/verify",
        JSON.stringify({ originalCode, confirmCode, forgottenDetails }),
        config
    );
    if (res.data.success) {
        return { success: true, msg: "" };
    }
    return { success: false, msg: "Code does not match" };
};

/**
 * Sends a username reminder and a temporary
 * password to a user who has forgotten their
 * login credentials. This email is only sent
 * after they have correctly entered the code
 * sent to them when they requested to change
 * their password
 *
 * @param {String} email -user's email
 */
export const sendTempCredentialsEmail = async (email) => {
    const res = await axios.post(
        "/api/email/send/reset",
        JSON.stringify({ email }),
        config
    );
    return res.data.success;
};
