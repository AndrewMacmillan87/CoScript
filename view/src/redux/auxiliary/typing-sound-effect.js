const typingSoundEffect = require("../../sound/320571__mlekoman__type-writing.wav");
const audioEffect = new Audio(typingSoundEffect);

/**
 * Starts playing the typewriter sound
 * effect
 */
export const playSoundEffect = () => {
    audioEffect.play();
};

/**
 * Stops playing the typewriter sound
 * effect
 */
export const stopSoundEffect = () => {
    audioEffect.pause();
    audioEffect.currentTime = 0;
};
