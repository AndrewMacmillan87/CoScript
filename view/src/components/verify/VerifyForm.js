import React, { Fragment } from "react";

import PostButton from "../buttons/PostButton";

import PropTypes from "prop-types";

const VerifyForm = ({ msg, name, placeholder, onSubmit, onChange }) => {
    return (
        <Fragment>
            <p className='verify-text'>{msg}</p>
            <div className='verify-email-container'>
                <form className='verify-email' onSubmit={e => onSubmit(e)}>
                    <input
                        className='v-email'
                        type='text'
                        name={name}
                        placeholder={placeholder}
                        onChange={e => onChange(e)}
                        onBlur={e => (e.target.placeholder = { placeholder })}
                        onFocus={e => (e.target.placeholder = "")}
                    />
                    <PostButton classes='square-btn' value='Submit' />
                </form>
            </div>
        </Fragment>
    );
};

VerifyForm.propTypes = {
    msg: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string.isRequired,
    onSubmit: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired
};

export default VerifyForm;
