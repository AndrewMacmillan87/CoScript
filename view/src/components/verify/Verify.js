import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";

import VerifyForm from "./VerifyForm";

import {
    sendEmail,
    verifyPWCode,
    sendTempCredentialsEmail,
} from "../../redux/auxiliary/sendEmail";

import "./verify.css";

export class Verify extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            userEmail: "",
            submitted: false,
            emailError: false,
            codeError: false,
            confirmCode: "",
            sending: false,
            codeSent: false,
        };
        this.onChange = this.onChange.bind(this);
        this.showEmailError = this.showEmailError.bind(this);
        this.showCodeError = this.showCodeError.bind(this);
        this.showSendingSpinner = this.showSendingSpinner.bind(this);
        this.removeSendingSpinner = this.removeSendingSpinner.bind(this);
        this.showCodeError = this.showCodeError.bind(this);
        this.onSubmitEmail = this.onSubmitEmail.bind(this);
        this.onSubmitCode = this.onSubmitCode.bind(this);
        this.showCodeSentMessage = this.showCodeSentMessage.bind(this);
        this.removeCodeSubmitMessage = this.removeCodeSubmitMessage.bind(this);
    }

    /**
     * Sets the browser tab title and scrolls to the
     * top of the page when the component mounts
     */
    componentDidMount() {
        document.title = "CoScript - Verify email";
        window.scrollTo(0, 0);
    }

    /**
     * Sets the <[e.target.name]> state property to the
     * value of the event target
     *
     * @param {Object} e - event object
     */
    onChange(e) {
        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    /**
     * Temporarily displays a message to the user if
     * the email they entered didn't match any records
     * in the database
     */
    showEmailError() {
        this.setState({
            emailError: true,
        });
        setTimeout(() => {
            this.setState({
                emailError: false,
            });
        }, 4000);
    }

    /**
     * Temporarily displays a message to the user if
     * the code they entered didn't match the verification
     * code they were sent in an email
     */
    showCodeError() {
        this.setState({
            codeError: true,
            originalCode: "",
        });
        setTimeout(() => {
            this.setState({
                codeError: false,
            });
        }, 4000);
    }

    /**
     * Sets the 'sending' state property to true to display
     * a loading spinner while an email is sent to the user
     * with a verification code
     */
    showSendingSpinner() {
        this.setState({
            sending: true,
        });
    }

    /**
     * Sets the 'sending' state property to false to remove
     * a loading spinner once an email has been sent to a
     * user with a verification code
     */
    removeSendingSpinner() {
        this.setState({
            sending: false,
        });
    }

    /**
     * Sets the 'codeSent' state property to true to display
     * a 'Please wait...' message to the user while the code
     * is verified
     */
    showCodeSentMessage() {
        this.setState({
            codeSent: true,
        });
    }

    /**
     * Sets the 'codeSent' state property to false to remove
     * a loading spinner once the code has been verified
     */
    removeCodeSubmitMessage() {
        this.setState({
            codeSent: false,
        });
    }

    /**
     * Shows an error message if the user didn't enter a
     * valid email. If the email is valid, the user is then
     * sent an email with a verification code
     *
     * @param {Object} e - event object
     */
    async onSubmitEmail(e) {
        e.preventDefault();
        this.showSendingSpinner();
        const { userEmail } = this.state;
        if (!userEmail) {
            this.showEmailError();
        } else {
            const subject = "User credentials reminder";
            const source = "reset";
            const res = await sendEmail(userEmail, subject, source);
            localStorage.setItem("originalCode", res.data.originalCode);
            this.setState({
                submitted: true,
            });
            this.removeSendingSpinner();
        }
    }

    /**
     * Firstly displays a message to the user confirming
     * the code has been sent to the server. If the code
     * matches, the user is sent an email with a reminder
     * of their username and a temporary password they
     * can log in with **THIS NEEDS REVISING**
     *
     * @param {Object} e - event object
     */
    async onSubmitCode(e) {
        e.preventDefault();
        this.showCodeSentMessage();
        const { userEmail, confirmCode } = this.state;
        const { history } = this.props;
        const originalCode = localStorage.getItem("originalCode");
        const res = await verifyPWCode({
            originalCode,
            confirmCode,
            forgottenDetails: true,
        });
        if (res.success) {
            await sendTempCredentialsEmail(userEmail);
            history["from"] = "verify";
            history.push("/redirect");
        } else {
            this.showCodeError();
            this.removeCodeSubmitMessage();
        }
    }

    render() {
        const { sending, codeSent } = this.state;
        const spinner = require("../../img/loading.svg");

        const { submitted, emailError, codeError } = this.state;
        return (
            <main className='verify-body'>
                <section className='verify-container'>
                    <h1 className='coscript-h1'>CoScript</h1>
                    <VerifyForm
                        msg='Please enter the email address you signed up
                                with:'
                        name='userEmail'
                        placeholder='Email'
                        onSubmit={this.onSubmitEmail}
                        onChange={this.onChange}
                    />
                    {submitted && (
                        <Fragment>
                            <VerifyForm
                                msg='Please enter the code we just sent to your email
                                address:'
                                name='confirmCode'
                                placeholder='Code'
                                onSubmit={this.onSubmitCode}
                                onChange={this.onChange}
                            />
                            {codeSent && (
                                <p className='code-sent-message'>
                                    Please wait...
                                </p>
                            )}
                        </Fragment>
                    )}
                    {emailError && (
                        <p className='verification-error'>
                            Email cannot be blank
                        </p>
                    )}
                    {codeError && (
                        <p className='verification-error'>Codes do not match</p>
                    )}
                    {sending && (
                        <div className='sending-credentials-email-spinner-container'>
                            <img
                                className='sending-credentials-email-spinner'
                                src={spinner}
                                alt='sending'
                            />
                        </div>
                    )}
                    <Link className='home-link' to='/'>
                        Back Home
                    </Link>
                </section>
            </main>
        );
    }
}

export default withRouter(Verify);
