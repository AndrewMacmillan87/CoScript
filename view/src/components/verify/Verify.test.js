import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Verify } from "./Verify";
import VerifyForm from "./VerifyForm";
import PostButton from "../buttons/PostButton";

import * as auxFuncs from "../../redux/auxiliary/sendEmail";

const originalSendEmail = auxFuncs.sendEmail;
const originalVerifyPWCode = auxFuncs.verifyPWCode;
const originalSendTempEmail = auxFuncs.sendTempCredentialsEmail;

describe("Verify component", () => {
    let props;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        auxFuncs.sendEmail = jest.fn();
        auxFuncs.verifyPWCode = jest.fn(function () {
            return { success: true };
        });
        auxFuncs.sendTempCredentialsEmail = jest.fn();
        props = {
            history: {
                from: "",
                push: jest.fn(),
            },
        };
    });

    afterEach(() => {
        auxFuncs.sendEmail = originalSendEmail;
        auxFuncs.verifyPWCode = originalVerifyPWCode;
        auxFuncs.sendTempCredentialsEmail = originalSendTempEmail;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Verify />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of verify-body", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(wrapper.find({ className: "verify-body" }).length).toEqual(1);
    });

    it("should render section element with a className of verify-container", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(wrapper.find({ className: "verify-container" }).length).toEqual(
            1
        );
    });

    it("should render the VerifyForm component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find(VerifyForm).length).toEqual(1);
    });

    it("should render a Link component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Link).length).toEqual(1);
    });

    it("should call `sendEmail()` when the form is submitted successfully", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "userEmail", value: "test@email.com" },
            });
        wrapper
            .find("form")
            .at(0)
            .simulate("submit", { preventDefault: jest.fn() });
        expect(auxFuncs.sendEmail).toBeCalled();
    });

    it("should call `verifyPWCode()` when the second form is submitted successfully", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify {...props} />
                </Router>
            </Provider>
        );
        wrapper.find(Verify).setState({ submitted: true });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "confirmCode", value: "123456" },
            });
        wrapper
            .find("form")
            .at(1)
            .simulate("submit", { preventDefault: jest.fn() });
        await expect(auxFuncs.verifyPWCode).toBeCalled();
        expect(auxFuncs.sendTempCredentialsEmail).toBeCalled();
    });
});

describe("VerifyForm component", () => {
    it("should render  a p element with a className of verify-text", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(1);
        expect(wrapper.find({ className: "verify-text" }).length).toEqual(1);
    });

    it("should render a form element with a className of verify-email", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find("form").length).toEqual(1);
        expect(wrapper.find({ className: "verify-email" }).length).toEqual(1);
    });

    it("should render 2 input elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find("input").length).toEqual(2);
    });

    it("should render the PostButton component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Verify />
                </Router>
            </Provider>
        );
        expect(wrapper.find(PostButton).length).toEqual(1);
    });
});
