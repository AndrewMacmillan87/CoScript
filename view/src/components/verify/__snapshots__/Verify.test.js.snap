// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify component renders correctly 1`] = `
<main
  className="verify-body"
>
  <section
    className="verify-container"
  >
    <h1
      className="coscript-h1"
    >
      CoScript
    </h1>
    <p
      className="verify-text"
    >
      Please enter the email address you signed up
 with:
    </p>
    <div
      className="verify-email-container"
    >
      <form
        className="verify-email"
        onSubmit={[Function]}
      >
        <input
          className="v-email"
          name="userEmail"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          type="text"
        />
        <input
          className="square-btn"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
    <a
      className="home-link"
      href="/"
      onClick={[Function]}
    >
      Back Home
    </a>
  </section>
</main>
`;
